accounts-db/src/accounts.rs

```diff
@@ -1260,7 +1345,271 @@ mod tests {
+            Hash::default(),
+            instructions,
+        );
+        let tx0 = new_sanitized_tx(&[&keypair1], message, Hash::default());
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair0.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx1 = new_sanitized_tx(&[&keypair0], message, Hash::default());
+        let txs2 = vec![tx0, tx1];
+        let (results2, self_conflicting_batch) =
+            accounts.lock_accounts(txs2.iter(), MAX_TX_ACCOUNT_LOCKS, true);
+
+        assert_eq!(self_conflicting_batch, false);
+        // the first tx conflicts with the previous batch
+        assert_eq!(
+            results2,
+            vec![
+                Err(TransactionError::AccountInUse), // outstanding write lock
+                Ok(()),                              // outstanding read lock, acquired read lock
+            ],
+        );
+        assert!(accounts
+            .account_locks
+            .lock()
+            .unwrap()
+            .write_locks
+            .contains_key(&keypair1.pubkey()));
+
+        accounts.unlock_accounts(txs.iter().zip(&results1));
+        accounts.unlock_accounts(txs2.iter().zip(&results2));
+
+        // batch write-read conflict. outstanding (read/write lock)
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair3.pubkey(), keypair0.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx = new_sanitized_tx(&[&keypair3], message, Hash::default());
+        let (results0, _) = accounts.lock_accounts([tx.clone()].iter(), MAX_TX_ACCOUNT_LOCKS, true);
+
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair2.pubkey(), keypair1.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx0 = new_sanitized_tx(&[&keypair2], message, Hash::default());
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx1 = new_sanitized_tx(&[&keypair1], message, Hash::default());
+        let txs = vec![tx0, tx1];
+        let (results1, _) = accounts.lock_accounts(txs.iter(), MAX_TX_ACCOUNT_LOCKS, true);
```

Comment:
add an assert on 2nd tuple item. This is new code added in this PR; we should be testing it everywhere that we are also asserting on the lock results.